### ISSUE TITLE
PayPal Vault Payment - Throws invalid payment options error occassionally

### ISSUE THREAD

**Issue by rtm619** [braintree/braintree-web#583]
### General information

* SDK version: 3.69.0
* Environment: Both Sandbox and Production
* Browser and OS 

### Issue description

Hi Team,

We have a PayPal checkout in our site and earlier we had the **PayPal Checkout** v3 version integrated into our site. Due to some requirements, we changed our configuration to **PayPal Vault** v3 version. 

After the changes, however, we noticed that some users are unable to open PayPal window due to an error which states that the PayPal payment options are invalid. We double checked our config and we're pretty confident that our integration was correctly done. 

What we also noticed was if we clear out the browser cache for the site, then this issue was getting resolved.

Hence, I would like to know if there's anything specific we need to take a look at, in order to make sure users do not encounter this error.

Thanks and regards,
Reetam

**Comment by crookedneighbor** [comment]
Can you provide your client side configuration?

**Comment by rtm619** [comment]
This is the paypal implementation function:

```
const createPaypalButton = async (
    paypalClientKey,
    variant,
    payPalPaymentOptions,
    handlePaymentSuccess,
    startLoading,
    stopLoading,
    validate
  ) => {
    try {
      braintree.client
        .create({
          authorization: paypalClientKey
        })
        .then((clientInstance) => {
          return braintree.paypalCheckout.create({
            client: clientInstance
          });
        })
        .then((paypalCheckoutInstance) => {
          return paypalCheckoutInstance.loadPayPalSDK({
            currency: payPalPaymentOptions?.currency,
            vault: true,
            intent: 'tokenize'
          });
        })
        .then((paypalCheckoutInstance) => {
          setTimeout(() => {
            const { paypal } = window;
            let paypalActions;
            return paypal
              .Buttons({
                fundingSource: paypal.FUNDING.PAYPAL,
                paypal: true,
                locale: 'en-CA',
                commit: false,
                style: {
                  color: 'gold', // settings.style,
                  label: 'checkout',
                  tagline: false,
                  shape: 'rect', // pill, rect
                  size: 'responsive' // device !== 'sm' ? 'large' : 'small', // small, medium, large, responsive
                },
  
                onInit(data, actions) {
                  const valid = validate(false);
                  if (valid) {
                    actions.enable();
                  } else {
                    actions.disable();
                  }
                  paypalActions = actions;
                },
  
                createBillingAgreement() {
                  return paypalCheckoutInstance.createPayment(payPalPaymentOptions);
                },
  
                onApprove(data) {
                  return paypalCheckoutInstance.tokenizePayment(data).then((payload) => {
                    handlePaymentSuccess(payload);
                  });
                },
  
                onClick() {
                  startLoading();
                  const valid = validate(true);
                  if (valid) {
                    paypalActions.enable();
                  }
                },
  
                validate(actions) {
                  console.log('validation phase', actions);
                },
  
                onCancel() {
                  stopLoading();
                },
  
                onError(err) {
                  stopLoading();
                  console.error('PayPal error', err);
                }
              })
              .render(`#paypal-button-${variant}`);
          }, 5);
        })
        .then(() => {
          // The PayPal button will be rendered in an html element with the ID
          // `paypal-button`. This function will be called when the PayPal button
          // is set up and ready to be used
        });
  
      return true;
    } catch (error) {
      return false;
    }
  };
```

This is the payment options object that is passed:
```
const payPalPaymentOptions = {
    amount: 33.05,
    enableShippingAddress: false,
    flow: 'vault',
    displayName: 'MY COMPANY',
    intent: 'tokenize',
    offerCredit: false,
    shippingAddressEditable: false,
    currency: 'CAD'
  };
```

**Comment by crookedneighbor** [comment]
Are any of the options in `payPalPaymentOptions` dynamic (other than `amount`, I assume)?

In the error, there should be a `details.originalError` message that should give more information about what the underlying problem is.

**Comment by rtm619** [comment]
@crookedneighbor Nope, aside from amount, all other values are static. I'll try to log the error message, but I'd have to find out how to reproduce it in the first place. Please let me know if there's anything specific we should look into.

**Comment by crookedneighbor** [comment]
Everything you posted looks like it _should_ work. I assume the error is coming from the `createPayment` call, which means the Braintree API believes the payment options are invalid, and if that is the case it should also tell you _what_ was invalid. (in that originalError property)

Is it possible the `amount` field is malformed somehow?

**Comment by rtm619** [comment]
@crookedneighbor Nope, it looks right to me. I'll keep you posted in case I reproduce it.

**Comment by rtm619** [comment]
@crookedneighbor We're sometimes using the `shippingAddressOverride` option to give a shipping address by default. If the `recipientName` is undefined, will PayPal throw this error?

**Comment by crookedneighbor** [comment]
A malformed shippingAddressOverride can certainly cause that error

**Comment by rtm619** [comment]
@crookedneighbor Cool, thanks for the confirmation. I will apply the fix. Thanks a lot!

**Comment by crookedneighbor** [comment]
I'm having issues using `shippingAddressOverride` at all in the vault flow. I'm going to consult with our PayPal API folks.

**Comment by rtm619** [comment]
@crookedneighbor Hmm, that's strange. Should I reopen this ticket?

**Comment by rtm619** [comment]
@crookedneighbor I redeployed my fix, and I see that the issue is still present. Please let me know if you get any update on this from the PayPal team.

**Comment by crookedneighbor** [comment]
Sorry for the delay, I'm in contact with our PayPal team on this to try and figure it out.

**Comment by crookedneighbor** [comment]
On my end, I was able to get it to work just fine as long as I made sure to set `enableShippingAddress = true` when using `shippingAddressOverride`. If I had it set to false, or didn't pass it at all, it conflicted with setting up the billing agreement with a shipping address override.

If you believe something else is going on, feel free to reach out to our support team with timestamps (or the debug id from the call to the `setup_billing_agreement` route) so we can check in our logs. https://developer.paypal.com/braintree/help
